File: Pandacap.Data/InboxArtworkDeviation.cs
namespace Pandacap.Data
{
    public class InboxArtworkDeviation : IPost, IPostImage
    {
        public Guid Id { get; set; }

        public Guid CreatedBy { get; set; }

        public string? Username { get; set; }
        public string? Usericon { get; set; }

        public DateTimeOffset Timestamp { get; set; }
        public bool MatureContent { get; set; }

        public string? Title { get; set; }
        public string? LinkUrl { get; set; }

        public class DeviantArtThumbnailRendition
        {
            public string Url { get; set; } = "";
            public int Width { get; set; }
            public int Height { get; set; }
        }

        public string? ThumbnailUrl { get; set; }

        public DateTimeOffset? DismissedAt { get; set; }

        string IPost.Id => $"{Id}";

        string IPost.DisplayTitle => Title ?? $"{Id}";

        IEnumerable<IPostImage> IPost.Images => [this];

        string? IPostImage.AltText => null;
    }
}

File: Pandacap.Data/DeviantArt/InboxArtworkDeviation.fs
namespace Pandacap.Data

open System

/// An artwork submission posted by a user who this instance's owner follows on DeviantArt.
type InboxArtworkDeviation() =
    member val Id = Guid.Empty with get, set
    member val CreatedBy = Guid.Empty with get, set
    member val Username = nullString with get, set
    member val Usericon = nullString with get, set
    member val Timestamp = DateTimeOffset.MinValue with get, set
    member val MatureContent = false with get, set
    member val Title = nullString with get, set
    member val LinkUrl = nullString with get, set
    member val ThumbnailUrl = nullString with get, set
    member val DismissedAt = nullDateTimeOffset with get, set

    interface IPost with
        member this.DisplayTitle = this.Title |> orString $"{this.Id}"
        member this.Id = $"{this.Id}"
        member this.Images = seq {{
            new IPostImage with
                member _.AltText = null
                member _.ThumbnailUrl = this.ThumbnailUrl
        }}
        member this.LinkUrl = this.LinkUrl
        member this.Timestamp = this.Timestamp
        member this.Usericon = this.Usericon
        member this.Username = this.Username
